pkg/commands/artifact/k8s.go

```diff
@@ -0,0 +1,274 @@
+)
+
+// K8sRun runs scan on kubernetes cluster
+func K8sRun(ctx *cli.Context) error {
+	opt, err := initOption(ctx)
+	if err != nil {
+		return xerrors.Errorf("option error: %w", err)
+	}
+
+	if err = log.InitLogger(opt.Debug, true); err != nil {
+		return err
+	}
+
+	cacheClient, err := initCache(opt)
+	if err != nil {
+		if errors.Is(err, errSkipScan) {
+			return nil
+		}
+		return xerrors.Errorf("cache error: %w", err)
+	}
+	defer cacheClient.Close()
+
+	// Disable DB update when using client/server
+	if opt.RemoteAddr == "" {
+		if err = initDB(opt); err != nil {
+			if errors.Is(err, errSkipScan) {
+				return nil
+			}
+			return xerrors.Errorf("DB error: %w", err)
+		}
+		defer db.Close()
+	}
+
+	cluster, err := k8s.GetCluster()
+	if err != nil {
+		return xerrors.Errorf("get k8s cluster: %w", err)
+	}
+
+	trivyk8s := trivyk8s.New(cluster).Namespace(opt.KubernetesOption.Namespace)
+
+	// list all kubernetes scannable artifacts
+	k8sArtifacts, err := trivyk8s.ListArtifacts(ctx.Context)
+	if err != nil {
+		return xerrors.Errorf("get k8s artifacts error: %w", err)
+	}
+
+	report, err := k8sRun(ctx, opt, cacheClient, k8sArtifacts)
+	if err != nil {
+		return xerrors.Errorf("k8s scan error: %w", err)
+	}
+	report.ClusterName = cluster.GetCurrentContext()
+
+	if err = k8sReport.Write(report, pkgReport.Option{
+		Format: "json", // for now json is the default
+		Output: opt.Output,
+	}); err != nil {
+		return xerrors.Errorf("unable to write results: %w", err)
+	}
+
+	exit(opt, report.Failed())
+
+	return nil
+}
+
+func k8sRun(cliContext *cli.Context, opt Option, cacheClient cache.Cache, k8sArtifacts []*artifacts.Artifact) (k8sReport.Report, error) {
+	ctx, cancel := context.WithTimeout(cliContext.Context, opt.Timeout)
+	defer cancel()
+
+	// progress bar
+	bar := pb.StartNew(len(k8sArtifacts))
+	if opt.NoProgress {
+		bar.SetWriter(io.Discard)
+	}
+	defer bar.Finish()
+
+	// image scanner configurations
+	imageScannerConfig, imageScannerOptions, err := initImageScannerConfig(ctx, opt, cacheClient)
+	if err != nil {
+		return k8sReport.Report{}, xerrors.Errorf("scanner config error: %w", err)
+	}
+
+	// config scanner configurations
+	configScannerConfig, configScannerOptions, err := initConfigScannerConfig(ctx, opt, cacheClient)
+	if err != nil {
+		return k8sReport.Report{}, xerrors.Errorf("scanner config error: %w", err)
+	}
+
+	vulns := make([]k8sReport.Resource, 0)
+	misconfigs := make([]k8sReport.Resource, 0)
+
+	// Loops once over all artifacts, and execute scanners as necessary. Not every artifacts has an image,
+	// so image scanner is not always executed.
+	for _, artifact := range k8sArtifacts {
+		bar.Increment()
+
+		// scan images if present
+		for _, image := range artifact.Images {
+			imageReport, err := k8sScan(ctx, image, imageScanner, imageScannerConfig, imageScannerOptions)
+			if err != nil {
+				// add error to report
+				log.Logger.Debugf("failed to scan image %s: %s", image, err)
```

Comment:
Any reason this is a debug message? I understand the error will be included in the report, but I feel like we should notify it in terminal.

Reply:
I asked a few friends to try it on their cluster with several resources (500-1000 resources), several resources were deployed with private images, and `trivy` was not configured to access the registry, the output becomes terrible, lots and lots of error logs, seems like everything is not working, and they cancelled the command, without waiting to generate a report , even to check at least the misconfigurations as the images were not reachable.

Reply:
Interesting. We probably should notify users if it is an authentication error like "To scan private images, please consider setting appropriate credentials." Let's keep it in our mind as a future enhancement.